core/io/resource.cpp

```diff
@@ -187,6 +187,72 @@ void Resource::reload_from_file() {
 	copy_from(s);
 }
 
+static void _dupe_sub_resources(const Variant &v, Node *p_for_scene, HashMap<Ref<Resource>, Ref<Resource>> &remap_cache) {
+	switch (v.get_type()) {
+		case Variant::ARRAY: {
+			Array a = v;
+			for (int i = 0; i < a.size(); i++) {
+				if (a[i].get_type() == Variant::ARRAY || a[i].get_type() == Variant::DICTIONARY) {
+					_dupe_sub_resources(a[i], p_for_scene, remap_cache);
+				} else if (a[i].get_type() == Variant::OBJECT) {
+					// Replace in array.
+					Ref<Resource> sr = a[i];
+					if (sr.is_valid() && sr->is_local_to_scene()) {
+						if (remap_cache.has(sr)) {
+							a[i] = remap_cache[sr];
+						} else {
+							Ref<Resource> dupe = sr->duplicate_for_local_scene(p_for_scene, remap_cache);
+							a[i] = dupe;
+							remap_cache[sr] = dupe;
+						}
+					}
+				}
+			}
+		} break;
+		case Variant::DICTIONARY: {
+			Dictionary d = v;
+			List<Variant> keys;
+			d.get_key_list(&keys);
+			for (const Variant &k : keys) {
+				if (k.get_type() == Variant::ARRAY || k.get_type() == Variant::DICTIONARY) {
+					_dupe_sub_resources(k, p_for_scene, remap_cache);
+				} else if (k.get_type() == Variant::OBJECT) {
```

Comment:
This can be another `case,` like in `_find_sub_resources()`, no?

Reply:
You're right, for arrays and for sub-resources this can be extracted.
Replacing a dictionary key however needs different logic, due to the `.erase()` call.